File: bindings/kotlin/scp-kt/src/main/kotlin/works/limn/scp/Economy.kt
// Economy.kt — Kotlin SDK economic amount-display surface.
//
// All monetary values in SCP are smallest-currency-unit integers on the wire
// (cents for USD, satoshis for BTC); the decimal scale used to render them is
// a purely SDK-side concern. This file holds that display surface and nothing
// else: pure integer/string arithmetic with no FFI dependency, mirrored by
// `format_amount` in the Python SDK and `formatAmount` in the TypeScript SDK.
//
// The `EconomyBindings` / `EconomyBridge` pair that used to live here was a
// zero-implementor stub over the deleted `CoroutineBridge` scaffold and was
// removed with it; it never reached the UniFFI economy exports.
//
// Provenance: spec §19 (Economic Governance), ADR-033, ADR-060.

package works.limn.scp

// ---------------------------------------------------------------------------
// Amount display formatting (ADR-060 SDK display surface)
// ---------------------------------------------------------------------------

/**
 * Number of decimal places for well-known currencies, keyed by uppercase
 * currency code. The SCP protocol does NOT store per-currency decimals -- the
 * wire form is always a smallest-unit integer -- so this table lives entirely
 * in the SDK for display. The same values are used across every SDK
 * (TypeScript, Python, Swift) for cross-binding consistency.
 */
private val KNOWN_CURRENCY_DECIMALS: Map<String, Int> =
    mapOf(
        "USD" to 2,
        "EUR" to 2,
        "GBP" to 2,
        "BTC" to 8,
        "SAT" to 0,
        "SOL" to 9,
        "USDC" to 6,
        "ETH" to 18,
    )

private fun formatWithDecimals(
    amount: ULong,
    decimals: Int,
): String {
    // Operate on the decimal digit string directly (no divisor arithmetic), so
    // any [decimals] -- even beyond a ULong's digit count -- formats exactly
    // with no overflow. A full-width ULong formats exactly.
    val digits = amount.toString()
    if (decimals == 0) {
        // The amount is already in whole display units -- no fraction.
        return digits
    }
    if (digits.length <= decimals) {
        return "0." + "0".repeat(decimals - digits.length) + digits
    }
    val split = digits.length - decimals
    return digits.substring(0, split) + "." + digits.substring(split)
}

/**
 * Formats a smallest-unit monetary amount as a human-readable decimal string,
 * applying the currency's decimal scale.
 *
 * Pure integer/string arithmetic (no floating point), so a full-width [ULong]
 * formats exactly.
 *
 * ```kotlin
 * formatAmount(150uL, "USD")        // "1.50"
 * formatAmount(100_000_000uL, "BTC") // "1.00000000"
 * ```
 *
 * @param amount Smallest-unit amount (e.g. cents, satoshis).
 * @param currency A known currency code (case-insensitive).
 * @return The human-decimal representation.
 * @throws IllegalArgumentException (SCP-ECON-12070) if the currency is unknown;
 *     use the [formatAmount] overload taking `decimals` for unknown/custom
 *     currencies. This is a pure SDK-side display helper that never touches the
 *     FFI bridge, so it raises an idiomatic argument exception rather than a
 *     [works.limn.scp.bridge.BridgeException] (which carries FFI error codes);
 *     the `SCP-ECON-12070` code is kept in the message for cross-SDK parity.
 */
fun formatAmount(
    amount: ULong,
    currency: String,
): String {
    val decimals =
        KNOWN_CURRENCY_DECIMALS[currency.uppercase()]
            ?: throw IllegalArgumentException(
                "[SCP-ECON-12070] unknown currency \"$currency\" has no known decimals; " +
                    "use formatAmount(amount, decimals) with an explicit scale",
            )
    return formatWithDecimals(amount, decimals)
}

/**
 * Formats a smallest-unit monetary amount using an explicit decimal scale, for
 * unknown or custom currencies.
 *
 * @param amount Smallest-unit amount (e.g. cents, satoshis).
 * @param decimals The number of fractional decimal places (0..100).
 * @return The human-decimal representation.
 * @throws IllegalArgumentException (SCP-ECON-12070) if [decimals] is out of
 *     range. Pure SDK-side display helper — no FFI bridge, so an idiomatic
 *     argument exception rather than a
 *     [works.limn.scp.bridge.BridgeException]; the `SCP-ECON-12070` code is
 *     kept in the message for cross-SDK parity.
 */
fun formatAmount(
    amount: ULong,
    decimals: Int,
): String {
    // `require` raises IllegalArgumentException — the idiomatic non-bridge
    // exception; the SCP-ECON-12070 code stays in the message for cross-SDK
    // parity.
    require(decimals in 0..100) { "[SCP-ECON-12070] decimals must be in 0..100, got $decimals" }
    return formatWithDecimals(amount, decimals)
}

File: bindings/typescript/src/economy.ts
/**
 * Economic governance types for the SCP TypeScript SDK.
 *
 * Defines wire types for cost estimation, budget tracking, and
 * antispam velocity. The functional entry points
 * (`estimateCost`, `budgetGrant`, `antispamVelocity`, etc.) moved onto
 * the {@link SCP} class in Phase 4 PR 4 (#1549, ADR-048) as
 * `scp.economyEstimateCost(...)`, `scp.economyBudgetGrant(...)`,
 * `scp.economyAntispamVelocity(...)` and so on. The free-function
 * shims that predated ADR-048 were deleted in the same commit.
 *
 * See spec section 19 (Economic Governance) and ADR-033.
 */

import { EconomyError } from "./errors";

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

/** Observable metrics for cost estimation and formula evaluation. */
export interface ObservableMetrics {
  /** Messages per minute in this context. */
  readonly contextMessageRate?: number;
  /** Current member count. */
  readonly memberCount?: number;
  /** Relay-level queue depth. */
  readonly relayQueueDepth?: number;
  /** UTC hour (0-23). */
  readonly timeOfDay?: number;
  /** Sender's messages in sliding window. */
  readonly senderVelocity?: number;
  /** Context storage usage in bytes. */
  readonly storageUsage?: number;
}

/** Paid action type for cost estimation. */
export type PaidActionType =
  | "MessageSend"
  | "OutletCall"
  | "ContextJoin"
  | "SubscriptionPeriod"
  | "ByteStored";

// ---------------------------------------------------------------------------
// Amount display formatting (ADR-060 SDK display surface)
// ---------------------------------------------------------------------------

/**
 * Number of decimal places for well-known currencies, keyed by uppercase
 * currency code. The SCP protocol does NOT store per-currency decimals — the
 * wire form is always a smallest-unit integer — so this table lives entirely
 * in the SDK for display purposes. The same values are used across every SDK
 * (Python, Swift, Kotlin) for cross-binding consistency.
 */
const KNOWN_CURRENCY_DECIMALS: Readonly<Record<string, number>> = {
  USD: 2,
  EUR: 2,
  GBP: 2,
  BTC: 8,
  SAT: 0,
  SOL: 9,
  USDC: 6,
  ETH: 18,
};

function formatWithDecimals(amount: bigint, decimals: number): string {
  if (amount < 0n) {
    throw new EconomyError(
      `amount must be non-negative, got ${amount.toString()}`,
      "SCP-ECON-12070",
    );
  }
  if (!Number.isInteger(decimals) || decimals < 0 || decimals > 100) {
    throw new EconomyError(
      `decimals must be an integer in 0..=100, got ${decimals}`,
      "SCP-ECON-12070",
    );
  }
  if (decimals === 0) {
    // The amount is already expressed in whole display units — no fraction.
    return amount.toString();
  }
  const divisor = 10n ** BigInt(decimals);
  const whole = amount / divisor;
  const fraction = amount % divisor;
  const fractionStr = fraction.toString().padStart(decimals, "0");
  return `${whole.toString()}.${fractionStr}`;
}

/**
 * Formats a smallest-unit monetary amount as a human-readable decimal string,
 * applying the currency's decimal scale.
 *
 * Purely integer/string arithmetic (no floating point), so amounts far beyond
 * `Number.MAX_SAFE_INTEGER` (2^53) format exactly.
 *
 * @example
 * formatAmount(150n, "USD"); // "1.50"
 * formatAmount(100_000_000n, "BTC"); // "1.00000000"
 * formatAmount(1500n, { decimals: 3 }); // "1.500"
 *
 * @param amount Smallest-unit amount (e.g. cents, satoshis). Must be non-negative.
 * @param currencyOrOptions A known currency code (case-insensitive) or an
 *   explicit `{ decimals }` override for unknown/custom currencies.
 * @throws {EconomyError} If the currency is unknown and no `decimals` override
 *   is supplied, or if `amount`/`decimals` are out of range.
 */
export function formatAmount(amount: bigint, currency: string): string;
export function formatAmount(amount: bigint, options: { decimals: number }): string;
export function formatAmount(
  amount: bigint,
  currencyOrOptions: string | { decimals: number },
): string {
  if (typeof currencyOrOptions === "object") {
    return formatWithDecimals(amount, currencyOrOptions.decimals);
  }
  const decimals = KNOWN_CURRENCY_DECIMALS[currencyOrOptions.toUpperCase()];
  if (decimals === undefined) {
    throw new EconomyError(
      `unknown currency ${JSON.stringify(currencyOrOptions)} has no known decimals; ` +
        "pass an explicit { decimals } override",
      "SCP-ECON-12070",
    );
  }
  return formatWithDecimals(amount, decimals);
}
